Adding Conditional to Status UI

jwshields opened the pull request:
Hello!

This PR is to add a conditional to the method which updates the status of threads.

Background:
I run CTW with a somewhat large number of threads. I have hundreds of threads that are updating the UI multiple times per second. As minimally intensive that a simple UI update may be, it still consumes processing power.
On an instance of CTW with a large number of threads, the time spent updating the UI may be liable to cause the responsiveness of the program to not be as immediate as it should be.
For example, my desktop machine has a Ryzen 1700x; CTW regularly sits around 1-4% CPU usage, and occasionally I have very small periods (less than a quarter of a second) where the UI does not respond immediately to clicks.

What:
The commit/PR adds in a conditional check which will only update the wait status of a thread once every 15 seconds if there is more than 10 minutes of waiting to do.
ex: "Waiting 1800 seconds" -> "Waiting 1785 seconds" -> "Waiting 1770 seconds"

Purpose:
The less the GUI needs to update the status of a thread means it spends less time wasting cycles by putting text up on the screen; For long lived threads, users (speaking from my point of view) don't truly need to see that there are another `"2581 seconds"` left before the next check, an exact number is not necessarily needed. As the thread gets closer to the time when it will actually be checked, the conditional will let it update the UI normally as time passes.

---

I've tested this build on my machine and it appeared to work without issue, does not seem that it introduced any other quirks or oddities.

SuperGouge committed:
Fixed downloading error when using a download speed limit (closes #80)

SuperGouge committed:
Fixed auto-update on supported environments

TLS 1.2 is not supported on .NET 2.0 runtimes but should be on .NET 3.5 or above if all the patches are installed. For more information: https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls#support-for-tls-12

SuperGouge committed:
Version 1.17.1: bug fixes

jwshields committed:
Merge branch 'master' of https://github.com/SuperGouge/ChanThreadWatch

jwshields committed:
Adding delay to update thread wait status if wait is greater than 15 minutes

jwshields committed:
Edit to 15m

jwshields committed:
Changing to 10m

SuperGouge commented:
Hello,

Thank you for the contribution and sorry for the delayed response.

Maybe instead of skipping updates and displaying "wrong" info, we could instead display a more vague time (e.g. "10 minutes")? Also, a setting to toggle this feature and/or set the threshold for this to trigger could be useful. What do you think?

jwshields commented:
Thank you for the suggestions. I will work on implementing those and update this PR-
Giving power to the users is a better idea overall; 👍

jwshields committed:
Add in option for changing threshold & enable/disable vague times

jwshields committed:
Change color, add tooltip

jwshields commented:
I've pushed a new commit to my fork with the requested changes.
This is not ready to merge just yet.

---

A few things to note:
1. Due to the size of the Settings window, I had to rearrange some of the items to accommodate for the new option. Please let me know if that's unsatisfactory or could be improved upon
2. I'm having an issue with the new option to switch between a vague time versus the old; See below
3. I was unsure of a good way to word the two new text areas, mostly `threshold of minutes before showing true time` - I believe `thread status format` gives a good indicator, but it could be improved upon - but maybe it would be worth not switching to seconds (or having the threshold) at all. I don't know.

Regarding item 2: 
From the settings menu, if I toggle between seconds and minutes, changing from minutes -> seconds will allow the thread to resume updating the UI as normal, but changing from seconds -> minutes does not update the thread until the next time period (ie, it leaves the seconds there until it hits another minute, at which that point it updates.
I could not figure out a way to trigger updating threads if there's a change in the settings area- If you wouldn't mind giving some guidance on how to force the status update if there's a change in the settings display, that'd be much appreciated. I didn't want to be overzealous and include an "update thread status" on the Okay button or anything of that nature.

---

Attaching a few screenshots:

Vague time disabled:
![image](https://user-images.githubusercontent.com/7400556/56187733-6eeb4d80-5fd8-11e9-9463-771fc31f56cf.png)

Vague time enabled & Thread above threshold:
![image](https://user-images.githubusercontent.com/7400556/56187721-68f56c80-5fd8-11e9-8588-e5d3f0ab45af.png)

Vague time enabled & Thread below threshold
![image](https://user-images.githubusercontent.com/7400556/56187736-714da780-5fd8-11e9-8429-be8819822a3d.png)

jwshields committed:
Fixing issues with simple status'

jwshields committed:
Adding additional file

jwshields commented:
Hello!

I've been a tad busy and just got around to updating things.
With the two commits I just pushed, I've fixed the issues I've listed in my previous comment;

>1. Due to the size of the Settings window, I had to rearrange some of the items to accommodate for the new option. Please let me know if that's unsatisfactory or could be improved upon

More rearrangement has occurred, I tried to clean things up a bit more.
I'd appreciate input on that; the asterisks correspond to the tip down at the bottom, as well as each having a tooltip popup.
The entire "thread status" box has a tooltip as well.
Please see the image attached at the bottom of this comment.

>I'm having an issue with the new option to switch between a vague time versus the old; See below

This has been fixed. I had to make some changes to the ThreadWatcher class and add in some checks to compare & see if settings had changed. Tested on my machine without issue.

>I was unsure of a good way to word the two new text areas

I've reworked the wording for this new feature;

---


Image1:
Settings window rearrangements
![image](https://user-images.githubusercontent.com/7400556/58144478-70291f00-7c03-11e9-90b0-da8ac242ab2b.png)

jwshields committed:
Forgot to remove this

SuperGouge commented on frmSettings.resx:
```diff
@@ -117,4 +117,28 @@
+  <data name="txtThreadStatusBoxThreshold.ToolTip" xml:space="preserve">
+    <value>Select the "Thread Status Threshold"
+The dropdown allows for always showing the number of seconds remaining in the thread status,
+or the number minutes remaining, until a specified threshold, at which seconds begin to be shown again.
```
```suggestion
or the number of minutes remaining, until a specified threshold, at which seconds begin to be shown again.
```

SuperGouge commented on frmSettings.resx:
```diff
@@ -117,4 +117,28 @@
+  <data name="ThreadStatusBox.ToolTip" xml:space="preserve">
+    <value>Select the "Thread Status Threshold"
+The dropdown allows for always showing the number of seconds remaining in the thread status,
+or the number minutes remaining, until a specified threshold, at which seconds begin to be shown again.
```
```suggestion
or the number of minutes remaining, until a specified threshold, at which seconds begin to be shown again.
```

SuperGouge commented on frmSettings.resx:
```diff
@@ -117,4 +117,28 @@
+  <data name="groupBox1.ToolTip" xml:space="preserve">
+    <value>Select the "Thread Status Threshold"
+The dropdown allows for always showing the number of seconds remaining in the thread status,
+or the number minutes remaining, until a specified threshold, at which seconds begin to be shown again.
```
```suggestion
or the number of minutes remaining, until a specified threshold, at which seconds begin to be shown again.
```

SuperGouge commented on Classes/ThreadWatcher.cs:
```diff
@@ -28,6 +28,7 @@ public class ThreadWatcher {
         private bool _oneTimeDownload;
         private int _checkIntervalSeconds;
         private int _minCheckIntervalSeconds;
+        private bool? _threadstatustype = Settings.ThreadStatusSimple;
```
```suggestion
        private bool? _threadStatusSimple = Settings.ThreadStatusSimple;
```

SuperGouge commented on Classes/ThreadWatcher.cs:
```diff
@@ -116,6 +117,11 @@ public class ThreadWatcher {
             }
         }
 
+        public bool? threadstatustype {
```
```suggestion
        public bool? ThreadStatusSimple {
```

SuperGouge commented on frmSettings.cs:
```diff
@@ -251,5 +257,22 @@ public partial class frmSettings : Form {
+
+        private void txtThreadStatusBoxThreshold_Leave(object sender, EventArgs e) {
+            long tmpThreadStatusThreshold;
+            if (!Int64.TryParse(txtThreadStatusBoxThreshold.Text, out tmpThreadStatusThreshold) || tmpThreadStatusThreshold < 0 || tmpThreadStatusThreshold > Int64.MaxValue)
```
I think `tmpThreadStatusThreshold` should be an `int` and parsed as such, the same way it is done in the `btnOK_Click` function. No need to check against the type's maximum value as that would have resulted in an incorrect parse.

jwshields replied:
Awesome, thanks - I had looked at the maximum speed text box and used the idea from there; this is cleaner

SuperGouge commented on frmSettings.cs:
```diff
@@ -40,6 +40,9 @@ public partial class frmSettings : Form {
             pnlBackupEvery.Enabled = chkBackupThreadList.Checked;
             txtBackupEvery.Text = (Settings.BackupEvery ?? 1).ToString();
+            ThreadStatusBox.SelectedIndex = Settings.ThreadStatusSimple == true ? 1 : 0;
+            txtThreadStatusBoxThreshold.Enabled = Settings.ThreadStatusSimple == true ? true : false;
```
```suggestion
            txtThreadStatusBoxThreshold.Enabled = Settings.ThreadStatusSimple == true;
```

SuperGouge commented on frmSettings.Designer.cs:
```diff
@@ -77,6 +81,8 @@ partial class frmSettings {
             this.chkCompletedFolder = new System.Windows.Forms.CheckBox();
             this.chkCompletedFolderRelative = new System.Windows.Forms.CheckBox();
+            this.frmSettingsToolTip = new System.Windows.Forms.ToolTip(this.components);
+            this.groupBox1 = new System.Windows.Forms.GroupBox();
```
Not of great importance, but can this have a more explicit name like the other group boxes? Something like `grpThreadStatus` or what you think is best.

jwshields replied:
Missed that one, thought I'd renamed it

SuperGouge commented on frmSettings.Designer.cs:
```diff
@@ -54,6 +56,9 @@ partial class frmSettings {
             this.grpGeneral = new System.Windows.Forms.GroupBox();
+            this.lblSupportedSitesNote = new System.Windows.Forms.Label();
+            this.txtThreadStatusBoxThreshold = new System.Windows.Forms.TextBox();
+            this.ThreadStatusBox = new System.Windows.Forms.ComboBox();
```
```suggestion
            this.cboThreadStatus = new System.Windows.Forms.ComboBox();
```
Just to follow the existing convention.

SuperGouge commented on frmSettings.cs:
```diff
@@ -251,5 +257,22 @@ public partial class frmSettings : Form {
                 General.GetAbsoluteDirectoryPath(path, Settings.ExeDirectory);
         }
+
+        private void ThreadStatusBox_SelectedIndexChanged(object sender, EventArgs e) {
```
```suggestion
        private void cboThreadStatus_SelectedIndexChanged(object sender, EventArgs e) {
```

SuperGouge commented on frmSettings.Designer.cs:
```diff
@@ -421,9 +433,46 @@ partial class frmSettings {
+            "Seconds (Default)",
+            "Minutes"});
+            this.ThreadStatusBox.Location = new System.Drawing.Point(6, 19);
+            this.ThreadStatusBox.Name = "ThreadStatusBox";
```
```suggestion
            this.cboThreadStatus.Name = "cboThreadStatus";
```

SuperGouge reviewed (changes requested):
Do you expect other types of thresholds to be added? I see the combo box has only seconds and minutes for now. If this is not expected to changed, maybe a checkbox would be a better fit? Conversely, if you expect more to be added, maybe the related setting should be changed from a boolean to an enum.

jwshields committed:
Adding requested changes

jwshields committed:
Converting to checkbox

jwshields committed:
Adding another condition to updates

jwshields commented:
I've made the requested changes & fixed the things you noted; also changed the setting to a check box, I don't personally see any other thresholds needed, possibly in the future it _might_ be wanted by somebody, but I don't believe it is important enough to incorporate more values right now.

jwshields committed:
Changing items to allow for edge cases when switching between thresholds & enabling/disabling feature where the status may not update properly

jwshields committed:
Fixed a bit with numbers and some odd updates

jwshields committed:
Split the string into a new method, edited some things to allow for a "0" threshold

jwshields committed:
Add param to download status to aid with sorting, reducing threads constantly switching positions when sorting on status

jwshields commented:
Sorry, not fully sure how reviewing with Github works - did not realize I was supposed to request a re-review

jwshields commented:
Closing this PR. It's been a while and I don't see much activity in this repo in the last few years.  
I've been maintaining my own sort of fork, feel free to take a look if you gain interest here again;  
Thank you for all the great & dedicated work you've done on this program
